cluster-backup-tool.md

```diff
@@ -0,0 +1,144 @@
+Cluster Backup Tool
+====================
+# Requirements
+## Hard
+- Ability to:
+    - Backup data to a file(s) at a given point in time
+    - Recreate data from archive
+- Must be backwards Compatible across x major versions
+- Supported in client/server mode
+
+## Optional/Future
+- Topology aware restoration
+    - Disabled by default, focus is on data
+
+- Import failure modes:
+    - Clusterwide: Rollback all on first error
+    - Per Cache: Only rollback importing of caches which encounter errors
+
+- Cache specific imports
+    - Greater flexibility
+    - Ability to retry failed cache imports
+    - Convert from one cache type to another
+
+- Configuration only backup/import
+
+- Enable "online" backups where the cluster creates a snapshot from a live cluster
+    - Utilise `CommitManager` as used by state transfer and conflict resolution
+    
+
+# Architecture
+The backup tool can be implemented as an extension to the CLI, with existing cache data stored as the corresponding CLI
+commands required to add the data to a cache. A cluster backup consists of a single global manifest which contains clusterwide
+metadata. As well as `.json` and `.dat` files for each cache and a single `counters.dat` file. These files are then
```

Comment:
I think so as counters are created via the CLI as follows:

```
create counter --initial-value=<val> --storage=<PERSISTENT|Volatile> --type=<strong|weak> <counter-name>
```
https://infinispan.org/docs/stable/titles/cli/cli.html#creating_counters

Reply:
I would "ignore" the volatile counters. 
And you will need a file to store the counter's configuration.
One possible issue it the CLI itself. It doesn't have an operation to set the counter's value. And weak counters can only do increments/decrements.

Reply:
> I would "ignore" the volatile counters.

I disagree as the purpose of the tool is to create a backup of the clusters current state so that it can be restored at a later date. We include non-persistent caches, so I don't see why we would do the same for counters?

> And you will need a file to store the counter's configuration.

That's the purpose of the `counters.dat` file, it's a batch file that recreates all of the counters with their initial configuration and current value.


> One possible issue it the CLI itself. It doesn't have an operation to set the counter's value. And weak counters can only do increments/decrements.

I was thinking that we can calculate the delta between the current value and the initial-value and then call:

```
[//containers/default/counters/my-weak-counter]> add --delta=2
```

It could also be possible to add `--delta` to the `create counter` command so that this can be achieved on a single line.

Reply:
> I disagree as the purpose of the tool is to create a backup of the clusters current state so that it can be restored at a later date. We include non-persistent caches, so I don't see why we would do the same for counters?

in my head, if a user configures a volatile counter, it is expected to be lost in case of shutdown, restarts, or crashes. 

I guess it wouldn't hurt to have them included :+1: